### PR TITLE
AirfRANS: gc=0.75 + T_max=5 extended — sweet-spot gradient clipping

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The gc sweep so far: gc=1.0 → 0.003904 (PR #2755), gc=0.5 → 0.00277 (PR #2774). Both diverge at ~ep205-208. The 28.9% improvement from gc=1.0 to gc=0.5 suggests an inverse relationship: lower clipping = deeper basins. But gc=0.5 may be too aggressive (diverges at similar epoch to gc=1.0, just with deeper basins).

**Hypothesis:** gc=0.75 sits between the two tested values and may find a sweet spot: deeper basins than gc=1.0 (0.003904) but more stable than gc=0.5 (0.00277, diverges ep205). If it survives longer (past ep210-220), the extended stability could push to similar or better depths than gc=0.5.

## Instructions

No code changes needed:

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 5 \
  --grad-clip 0.75 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --agent tanjiro \
  --wandb-name "tanjiro/airfrans-4L256d-gc075-extended" \
  --wandb-group "airfrans-gc-sweep"
```

**Key things to report:**
1. Best val_primary/surface_mse and at which epoch
2. What epoch does divergence occur (if any)?
3. Trough envelope comparison vs gc=0.5 run (0pt769m4) and gc=1.0 run (stxm16tv)
4. Grad norm behavior — is it between gc=0.5 and gc=1.0?

## Baseline

- **Current best:** val_primary/surface_mse = **0.00277** at ep150 (PR #2774, gc=0.5)
- **Reference gc=1.0:** 0.003904 at ep201 (PR #2755)
- **gc sweep target:** if gc=0.75 achieves < 0.00277 with later divergence, it wins
- **Reproduce (gc=0.5 winner):** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### gc=0.75 Extended Run (180-min budget)

| Metric | gc=0.75 (this run) | gc=0.5 (0pt769m4) | gc=1.0 (stxm16tv) |
|--------|--------------------|--------------------|---------------------|
| **Best val_primary/surface_mse** | **0.004467** (ep79) | **0.00277** (ep150) | **0.003904** (ep201) |
| Divergence epoch | **ep122** | ~ep205 | ~ep208 |
| Sub-0.005 troughs | 2 (ep45, ep79) | Multiple (ep77-204) | Sporadic (ep195-201) |
| Grad norm before divergence | 3.6–12.6 | — | — |
| Total epochs | 221 | 221 | 223 |

**Surface MSE breakdown at best (ep79):** Ux=1.03e-04, Uy=1.04e-05, nut=6.89e-06, p=0.01775
**full_val/volume_mse at best:** 0.01963

### What happened

**The sweet-spot hypothesis is falsified.** gc=0.75 is the worst of the three tested values:

1. **Shallower basins than both gc=0.5 and gc=1.0:** Best of 0.004467 vs 0.00277 (gc=0.5) and 0.003904 (gc=1.0). Never reached sub-0.004 territory.
2. **Much earlier divergence:** Catastrophic gradient explosion at ep119-122 (grad norms: 43 → 106 → 212 → ∞). This is 80+ epochs earlier than gc=0.5 or gc=1.0 (~ep205-208).
3. **Non-monotonic relationship:** The gc sweep shows gc=0.5 > gc=1.0 > gc=0.75. gc=0.75 appears to sit in a "worst of both worlds" zone — not aggressive enough clipping to reach the deep basins that gc=0.5 finds, but not loose enough to maintain the stability that gc=1.0 provides.

The trough envelope was still oscillating productively through ep40-113 with ~26 sub-0.01 readings, but the model never consolidated into consistently deep basins. The divergence cascade was rapid: ep118 (surface_mse=0.015, grad_norm=20) → ep119 (0.054, 44) → ep120 (0.100, 106) → ep121 (0.157, 212) → ep122 (2.375, ∞).

### gc sweep summary

| gc value | Best surface_mse | Best epoch | Divergence | Rank |
|----------|-----------------|------------|------------|------|
| 0.5 | 0.00277 | 150 | ~ep205 | 🥇 |
| 1.0 | 0.003904 | 201 | ~ep208 | 🥈 |
| 0.75 | 0.004467 | 79 | ep122 | 🥉 |

### Experiment details

- **W&B run:** ixvgal1f
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 0.75 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent tanjiro --wandb-name "tanjiro/airfrans-4L256d-gc075-extended" --wandb-group "airfrans-gc-sweep"`
- **Peak memory:** ~98GB GPU (H100), run completed without OOM
- **Runtime:** ~180 min (221 epochs, ~0.81 min/epoch)

### Suggested follow-ups

1. **gc=0.25 or gc=0.3** — the non-monotonic pattern suggests the optimal gc might be below 0.5. More aggressive clipping might push to even deeper basins, though divergence timing is hard to predict.
2. **gc=0.5 + T_max=10** — the advisor already noted this direction. Longer cosine cycles might delay divergence past ep205 while preserving gc=0.5's deep-basin finding ability.
3. **Warmup clipping schedule** — start with gc=1.0 for stability in early epochs, then anneal to gc=0.5 after ep100. This could combine gc=1.0's late-training stability with gc=0.5's basin depth.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2819 because ADVISOR: Closing — your gc=0.75 result of 0.004467 is definitively the worst of the gradient clipping sweep, well above the 0.001479 baseline. This confirms gc=0.5 is the optimal clipping value. Clear result, tanjiro — thanks for anchoring this boundary..
